Sources/CoreMetrics/SystemMetrics.swift

```diff
@@ -0,0 +1,98 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Metrics API open source project
+//
+// Copyright (c) 2018-2019 Apple Inc. and the Swift Metrics API project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of Swift Metrics API project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+
+public enum SystemMetricsProvider {
+    fileprivate static let lock = ReadWriteLock()
+    fileprivate static let queue = DispatchQueue(label: "com.apple.CoreMetrics.SystemMetricsProvider")
+    fileprivate static var shouldRunSystemMetrics: Bool = false
+
+    public static func bootstrapSystemMetrics() {
+        self.lock.withWriterLockVoid {
+            self.shouldRunSystemMetrics = true
+        }
+        DispatchQueue.global(qos: .background).async {
+            updateSystemMetrics()
+        }
+    }
+    
+    public static func cancelSystemMetrics() {
+        self.lock.withWriterLockVoid {
+            self.shouldRunSystemMetrics = false
+        }
+    }
+    
+    private enum SystemMetricsError: Error {
+        case FileNotFound
+    }
+    
+    fileprivate static func updateSystemMetrics() {
+        self.queue.asyncAfter(deadline: .now() + .seconds(2)) {
+            let shouldReturn = self.lock.withReaderLock { () -> Bool in
+                if !self.shouldRunSystemMetrics {
+                    return true
+                }
+                return false
+            }
+            if shouldReturn { return }
+            let prefix = "process_"
+            let pid = ProcessInfo.processInfo.processIdentifier
+            let ticks = Int32(_SC_CLK_TCK)
+            #if os(Linux)
```

Comment:
I suggest breaking this up to two discrete functions and then calling those depending on operating system. both should return some "model" object that can then be  used to record the metrics consistently across the operating systems